### ISSUE TITLE
[emacs 25] Some sync command hangs in latest emacs

### ISSUE THREAD

**Issue by gambhiro** [org-trello/org-trello#230]
Awesome project!

When I first installed `org-trello`, my emacs was installed from the [ubuntu emacs list](https://launchpad.net/~ubuntu-elisp/+archive/ubuntu/ppa) snapshots, which now installs emacs 25.

I could complete the setup process (consumer key and token), but when calling org-trello commands to sync, it would most often just hang, and I'd have to kill and restart emacs. Very strange, occasionally it would work, but I couldn't find out when.

Then I installed emacs 24.4.1 from source and now it works well.

**Comment by ardumont** [comment]
> Awesome project!
>

Thanks!
Glad to hear that :D

> When I first installed org-trello, my emacs was installed from the ubuntu emacs list snapshots, which now installs emacs 25.

Ok.

> I could complete the setup process (consumer key and token),

Yes, ok, it's because, this is mostly the browser that does everything.
Emacs just triggers some trello url and asks nicely the browser to open them and the user does the rest :D

> but when calling org-trello commands to sync, it would most often just hang, and I'd have to kill and restart emacs.

No error message of any kind?
Do you have some *Messages* buffer log for example?

> Very strange, occasionally it would work, but I couldn't find out when.

On the top of my head, I read recently that [Lars Magne Ingebrigtsen added the network security manager in emacs 's latest code](http://emacs.1067599.n5.nabble.com/Network-security-manager-td338799.html).
In this thread, breaking some tiers-code was a possibility. So, it might be one of the reason.

> Then I installed emacs 24.4.1 from source and now it works well.

Ok.
I indeed only tested with emacs that I could install:
- 24.3 (which was the default release when I started developing org-trello)
- 24.4 (I use 24.4 dayly)

running GNU/Linux distributions (currently NixOS).

So thanks for raising this warning.

Do you know, by any luck, when emacs 25 will be released?

**Comment by gambhiro** [comment]
No error messages, the status bar prints

> org-trello - Synchronizing the trello board 'Learning Something' to the org-mode file...

And then emacs gets stuck, as if waiting for some function to return. Curious, but probably most folks stay on emacs 24.4 anyway.

**Comment by ardumont** [comment]
> No error messages, the status bar prints
> 
> org-trello - Synchronizing the trello board 'Learning Something' to the org-mode file...
> 
> And then emacs gets stuck, as if waiting for some function to return. 

Thanks.

This looks like some process might wait for the user's input.
So my initial proposal seems right :D 

Anyway, I will try to install emacs 25 as soon as I have some time to try.

> Curious, but probably most folks stay on emacs 24.4 anyway.

It was [released on the 20th of October 2014](http://www.gnu.org/software/emacs/history.html) so it's not that huge of a gap :D

Cheers,

**Comment by kmartino** [comment]
I run Emacs 25.0.50.1 on a Mac (GNU Emacs 25.0.50.1 (x86_64-apple-darwin13.4.0, NS appkit-1265.21 Version 10.9.5 (Build 13F34)) of 2014-10-20) and can reproduce this. 

Quitting the command frees up whatever is blocking and it then completes. 

Thanks.

**Comment by ardumont** [comment]
I did not have time to look at this one yet.

> I run Emacs 25.0.50.1 on a Mac (GNU Emacs 25.0.50.1 (x86_64-apple-darwin13.4.0, NS appkit-1265.21 Version 10.9.5 (Build 13F34)) of 2014-10-20) and can reproduce this.

Ok.

Nothing is prompting you for anything by any chance?

Is it possible for you to copy/paste a sample of your *Message* buffer?

> Quitting the command frees up whatever is blocking and it then completes.

You quit using <kbd>C-g</kbd>?

> Thanks.

Thank you

Cheers,

**Comment by kmartino** [comment]
Hi,

I don't see anything prompting me, however, something could. 

Messages buffer output after starting to sync:

```
org-trello - Request 'sync org buffer from trello board'...

Loading /Users/myuser/.trello/config.el (source)...done

org-trello - Synchronizing the trello board 'Karl's Trello' to the org-mode file...
```

I hit `C-g` to quit and the Messages buffer has the following:

```
Quit

org-trello - Sync buffer from trello - Catch error: (wrong-type-argument hash-table-p wrong-type-argument)
```

**Comment by ardumont** [comment]
Hello,

I can reproduce and it seems not at all related to emacs 25's internals.

```log
...
org-trello - Archived trello-cards: (wrong-type-argument arrayp nil)
org-trello - Sync buffer from trello - Catch error: (wrong-type-argument hash-table-p wrong-type-argument)
....
```

From what I gather, there is some http request which does get a 404, thus returning nil, thus breaking somewhere on deferred action...
[As this part is executed in parallel](https://github.com/org-trello/org-trello/blob/master/org-trello-controller.el#L206-L222), this seems to block all results from continuing and is indeed unblock by the user's quit command.

The pb seems to be related to the archive cards request.

I'm still investigating the issue.

--

I keep forgetting mentioning this but to have more logs, you can add this in a buffer and evaluate it:

```lisp
(custom-set-variables '(*orgtrello-log/level* *OT/TRACE*))
```
The *Message* buffer will have then a lot more details to help pinpoint the pb.
(I need to add this in the [issue page](http://org-trello.github.io/contributions.html#issues).)

Cheers,

**Comment by ardumont** [comment]
Hello,

I made some small progress on the analysis...

> but when calling org-trello commands to sync, it would most often just hang,

It would help to narrow the issue by describing which other sync command hang.
Thanks.

For me, I reproduce with the `C-u M-x org-trello/sync-buffer`.
Is there more?

> and I'd have to kill and restart emacs. Very strange, occasionally it would work, but I couldn't find out when.

As suggested by @kmartino, <kbd>C-g</kbd> should be enough.

--

For what I gather, the problem lies finally with the parallel requests execution (one to retrieve archive cards and one for open cards) - https://github.com/org-trello/org-trello/blob/master/org-trello-controller.el#L206-L216.

The parallel command's contract is to block until it retrieves all the data from the multiple functions it has to execute in parallel before going to the next step (which is the buffer sync in our case).

But I don't understand why this blocks because the individual execution of those requests works as expected.

(Also, note that this behavior is only from emacs-25...).

I have adapted some code to execute sequentially and then the problem disappear.
This is less elegant but at least this works.

```emacs-lisp
(defun orgtrello-controller/do-sync-buffer-from-trello! ()
  "Full org-mode file synchronisation. Beware, this will block emacs as the request is synchronous."
  (lexical-let ((buffer-name (current-buffer))
                (board-name  (orgtrello-buffer/board-name!))
                (point-start (point))
                (board-id (orgtrello-buffer/board-id!)))
    (orgtrello-log/msg *OT/INFO* "Synchronizing the trello board '%s' to the org-mode file..." board-name)
    (deferred:$ ;; in emacs-25, deferred:parallel blocks in this particular context and I don't understand why so I retrieve sequentially for the moment. People, feel free to help and improve.
      (deferred:next
        (lambda ()
          (-> board-id
              orgtrello-api/get-archived-cards
              (orgtrello-query/http-trello 'sync)
              list)))
      (deferred:nextc it
        (lambda (cards)
          (-> board-id
              orgtrello-api/get-full-cards
              (orgtrello-query/http-trello 'sync)
              (cons cards))))
      (deferred:nextc it
        (lambda (trello-opened-and-archived-cards)
          (orgtrello-log/msg *OT/DEBUG* "Opened and archived trello-cards: %S" trello-opened-and-archived-cards)
          (let ((trello-cards          (car  trello-opened-and-archived-cards))
                (trello-archived-cards (cadr trello-opened-and-archived-cards)))
            ;; first archive the cards that needs to be
            (orgtrello-log/msg *OT/DEBUG* "Archived trello-cards: %S" trello-archived-cards)
            (orgtrello-buffer/archive-cards! trello-archived-cards)
            ;; Then update the buffer with the other opened trello cards
            (orgtrello-log/msg *OT/DEBUG* "Opened trello-cards: %S" trello-cards)
            (->> trello-cards
                 (mapcar 'orgtrello-data/to-org-trello-card)
                 (orgtrello-controller/sync-buffer-with-trello-cards! buffer-name)))))p
      (deferred:nextc it
        (lambda ()
          (orgtrello-buffer/save-buffer buffer-name)
          (goto-char point-start)
          (orgtrello-log/msg *OT/INFO* "Synchronizing the trello board '%s' to the org-mode file '%s' done!" board-name buffer-name)))
      (deferred:error it
        (lambda (err) (orgtrello-log/msg *OT/ERROR* "Sync buffer from trello - Catch error: %S" err))))))
```

If you are interested, you could add this code in a scratch buffer and eval this (<kbd>M-x eval-buffer</kbd>).
This will monkey-patch (boo!) your org-trello install to try and test on your end.

Thanks for getting back to me.

Cheers,

**Comment by kmartino** [comment]
Hi,
This works, however there is a puzzling message outtputted at the very beginning of the process that you see fly by in the mini-buffer and can be read in *Messages*

**Comment by ardumont** [comment]
Great news.

There is a typo inside thé code i sent you.
Remove the useless 'p' and all will be good.

Cheers,

**Comment by ardumont** [comment]
Here is the right code without typo:

```lisp
(defun orgtrello-controller/do-sync-buffer-from-trello! ()
  "Full org-mode file synchronisation. Beware, this will block emacs as the request is synchronous."
  (lexical-let ((buffer-name (current-buffer))
                (board-name  (orgtrello-buffer/board-name!))
                (point-start (point))
                (board-id (orgtrello-buffer/board-id!)))
    (orgtrello-log/msg *OT/INFO* "Synchronizing the trello board '%s' to the org-mode file..." board-name)
    (deferred:$ ;; in emacs-25, deferred:parallel blocks in this particular context and I don't understand why so I retrieve sequentially for the moment. People, feel free to help and improve.
      (deferred:next
        (lambda ()
          (-> board-id
              orgtrello-api/get-archived-cards
              (orgtrello-query/http-trello 'sync)
              list)))
      (deferred:nextc it
        (lambda (cards)
          (-> board-id
              orgtrello-api/get-full-cards
              (orgtrello-query/http-trello 'sync)
              (cons cards))))
      (deferred:nextc it
        (lambda (trello-opened-and-archived-cards)
          (orgtrello-log/msg *OT/DEBUG* "Opened and archived trello-cards: %S" trello-opened-and-archived-cards)
          (let ((trello-cards          (car  trello-opened-and-archived-cards))
                (trello-archived-cards (cadr trello-opened-and-archived-cards)))
            ;; first archive the cards that needs to be
            (orgtrello-log/msg *OT/DEBUG* "Archived trello-cards: %S" trello-archived-cards)
            (orgtrello-buffer/archive-cards! trello-archived-cards)
            ;; Then update the buffer with the other opened trello cards
            (orgtrello-log/msg *OT/DEBUG* "Opened trello-cards: %S" trello-cards)
            (->> trello-cards
                 (mapcar 'orgtrello-data/to-org-trello-card)
                 (orgtrello-controller/sync-buffer-with-trello-cards! buffer-name)))))
      (deferred:nextc it
        (lambda ()
          (orgtrello-buffer/save-buffer buffer-name)
          (goto-char point-start)
          (orgtrello-log/msg *OT/INFO* "Synchronizing the trello board '%s' to the org-mode file '%s' done!" board-name buffer-name)))
      (deferred:error it
        (lambda (err) (orgtrello-log/msg *OT/ERROR* "Sync buffer from trello - Catch error: %S" err))))))
```

Cheers,

**Comment by ardumont** [comment]
> Thanks for working thru this. org-trello looks to be very useful for me and I'd be glad
to learn more about its development to help.

That's good news :D

**Comment by kmartino** [comment]
Hi, unfortunately, a new problem seems to have developed (and I don't know if it is related yet):

The same operation above, when succeeding in bringing down cards from Trello onto my org-mode file, indents cards in a pattern I haven't yet figured out.

It does that immediately after the import. And then leaves the org-mode file in a weird state where any time I add new headline/element, items following get automatically indented.

**Comment by kmartino** [comment]
Ah, this seems related to comments. If I have a card with a comment, then when sync is ran, the org file gets malformed and in a state where if you add an item, items beneath get automatically indented.

**Comment by ardumont** [comment]
Ah yes...
I'm working on it on #232.

**Comment by kmartino** [comment]
Hi, 

I can confirm the new release solves this!

Thanks!

Karl

**Comment by ardumont** [comment]
Hello,

Thanks for the feedback Karl.

Cheers,

tony / @ardumont

**Comment by gambhiro** [comment]
Yes, I checked out the latest from melpa, and this works great now. I'm sorry I haven't been able to be more active. Many thanks!

**Comment by ardumont** [comment]
> Yes, I checked out the latest from melpa, and this works great now. 

Cool, thanks for the feedback.

> I'm sorry I haven't been able to be more active. 

Not a problem, Karl took over :D 

> Many thanks!

Thanks for raising this up in the first place!

Cheers,
